src/v3/src/util/widgetHooks.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { IdxStatus, IdxTransaction } from '@okta/okta-auth-js';
+
+import { HookFunction, HooksOptions, HookType } from '../../../types';
+import { getFormNameForTransaction } from './getEventContext';
+
+export class WidgetHooks {
+  private hooks: Map<string, Map<string, HookFunction[]>>;
+
+  /* eslint-disable no-restricted-syntax */
+  constructor(hooksOptions?: HooksOptions) {
+    this.hooks = new Map();
+    if (hooksOptions) {
+      for (const [formName, formHooks] of Object.entries(hooksOptions)) {
+        const formHooksMap = new Map<string, HookFunction[]>();
+        for (const [hookType, hooksByType] of Object.entries(formHooks)) {
+          if (hooksByType) {
+            formHooksMap.set(hookType, hooksByType);
+          }
+        }
+        this.hooks.set(formName, formHooksMap);
+      }
+    }
+  }
+
+  public addHook(hookType: HookType, formName: string, hook: HookFunction): void {
+    const formHooks = this.hooks.get(formName) || new Map<string, HookFunction[]>();
+    const hooksByType = formHooks.get(hookType) || [];
+    hooksByType.push(hook);
+    formHooks.set(hookType, hooksByType);
+    this.hooks.set(formName, formHooks);
+  }
+
+  /* eslint-disable no-await-in-loop */
+  public async callHooks(
+    hookType: HookType,
+    idxTransaction?: IdxTransaction,
+    error?: Error | null,
+  ): Promise<void> {
+    if (!(idxTransaction || error)) {
+      return;
+    }
```

Comment:
I think it's better to not pass `error` at all because its value is not used in method body.
```
callHooks(
    hookType: HookType,
    idxTransaction?: IdxTransaction | null
  )
```
If `idxTransaction` is null, then it's error case and need to execute hooks for "terminal"